### DISCUSSION TITLE
Walkthrough: using Typescript and imports with pb_hooks

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3341] [Show and tell]
I spent the day trying some different solutions to using pb_hooks in https://pockethost.io. We have some server-side functions that in theory could be moved to hooks, but I found it difficult to manage package dependencies and I really wanted full Typescript support too.

Here's what I came up with. Let me know what you guys think!

## Folder structure

Create `./src/hooks/*.pb.ts` files. These are your entry points. I recommend making many of them and keeping them focused on specific tasks. While you could make one `onRecordAfterUpdateRequest` to hold the ocean, it's better to use many purpose-built `onRecordAfterUpdateRequest` handlers. So make one per `.pb.ts` file.

```
pb_hooks/
src/
   hooks/
      some-purposeful-name.pb.ts
```

Example:

```ts
// src/hooks/some-sensible-name.pb.ts

/// <reference path="/path/to/pb_data/types.d.ts" />

onRecordAfterUpdateRequest((e) => {
  console.log(JSON.stringify(e.record)))
}, 'users')
```

## Build step

```bash
yarn add -D chokidar tsup
```

In `package.json`:

```json
  "scripts": {
    "build:hooks": "tsup ./src/hooks/*.ts -d ./pb_hooks",
    "dev:hooks": "chokidar './src/hooks/**' -c 'yarn build:hooks' --initial",
  }
```

During development, run `yarn dev:hooks` to watch for changes to `src/hooks` and compile them into `pb_hooks`.

## Running PocketBase with your generated hooks

```bash
pocketbase --hooksDir=/path/to/pb_hooks
```

## Including libraries, modules, etc.

Add `./src/hooks/lib.ts`. This file will serve as the 'kitchen sink' library entry point. You can include all your goodies in it just like you always do. Keep in mind, you'll need to `require()` this in your hook files, so be aware of how large this library grows. It will build as `./pb_hooks/lib.js` which is not autoloaded by PocketBase.

Some Typescript kung fu makes this work by creating a type that is the entire interface to your library:

```ts
type Lib = typeof import('./lib')
```

Using that, you can then require your library and cast it to the `Lib` type. This works in VSCode, too, even though there is no `.js` file!

```ts
const lib = require(`${__hooks}/lib.js`) as Lib
```

Putting it all together:

```ts
// ./src/hooks/lib.ts example

export const hello = ()=>console.log(`hello from library!`)
```

To use it in a hook, do this:

```ts
// src/hooks/some-sensible-name.pb.ts

type Lib = typeof import('./lib')

onRecordAfterUpdateRequest((e) => {
  const lib = require(`${__hooks}/lib.js`) as Lib
  const { hello } = lib
  hello()
}, 'users')

```

## Special note about assertion functions

If you use Typescript assertion functions, you need to know that they require an explicit type annotation. 

Suppose you have

```ts
// assert.ts

export function assert<T>(
  v: T | undefined | void | null,
  msg?: string,
): asserts v is T {
  if (!v) {
    throw new Error(msg || `Assertion failure`)
  }
}
```

In `lib.ts`, you'll want to do

```ts
import { assert } from './assert'
export type Assert = typeof assert
export { assert }
```

And in your hook files (`*.pb.ts`), it'll look like this

```ts

import { Assert } from './lib'

type Lib = typeof import('./lib')

onRecordAfterUpdateRequest((e) => {
  const lib = require(`${__hooks}/lib.js`) as Lib
  const assert: Assert = lib.assert
  assert(record, `Expected record here`)
}, 'users')
```

**Comment by einarpersson** [comment]
@benallfree Now that some time has passed, has this setup aged well or have it been reverted? (or evolved?)

I am interested in getting typescript support in hooks, but don't want to start go down a road that will cause more headache than it is worth.

---
On a side note: It would really be nice to not have to add an extra npm script-process, but rather have it integrated with whatever you are using already (in my case sveltekit/vite), and add a custom `tsconfig.json` in the hooks-folder that overrides necessary options. I am not proficient enough to know how to do it though. Have you experimented with such an approach?

**Reply by benallfree** [reply]
I ended up dropping TS because it became too unwieldy https://github.com/pockethost/pockethost/tree/master/src/mothership-app/pb_hooks/src. A lot of that decision came from working with pb_hooks and realizing they are still somewhat brittle. Errors are very difficult to troubleshoot (sometimes it just says `module error` with no further information or clues), and support for npm packages is very limited because most npm packages are not pure es5 and have at least some dependency on web or node APIs. All of this caused me to fall back to pure es5 and to keep modules very simple.

**Reply by Leftium** [reply]
@benallfree: Thanks for this guide and [ts-pb-hooks-starter](https://github.com/benallfree/ts-pb-hooks-starter)

I used your starter repo as the basis for the `pb` half of my [spock-stack-spa-starter](https://github.com/Leftium/spock-stack-spa-starter).

Your starter gave me a lot ideas and was a good start, but it actually had several minor copy-paste errors that prevented it from working out of the gate. I also added a few QOL updates. Would you care for a PR?

So far I haven't made any complex hooks, but there are zero TS errors.

**Reply by benallfree** [reply]
Thank you! Yes that repo is pretty old, if you want a good build approach look at [how pocodex does it](https://github.com/benallfree/pocodex/blob/main/tsup.config.ts).

**Reply by Leftium** [reply]
Wow! So you picked up TS again...

I think I've applied your pocodex TS build method (minus packaging/bun build stuff):
- https://github.com/Leftium/spock-stack-spa-starter/commit/7ac581986811825ff10a86a37d8bd775ce9d6a8f
- Anything I missed?
- Did you know [tsup has a watch mode](https://tsup.egoist.dev/#watch-mode)?
    - I tried using it, but it watched the wrong files (because files specified via CLI args? Or due to Windows?)
    - Could you quick test if it works when the entries are specified via tsup.config.ts?

**These new developments opened up a few interesting questions:**
1. Would it be better to integrate pocodex plugins in my spock-stack-spa-starter (instead of directly commiting/compiling [code like this](https://github.com/Leftium/spock-stack-spa-starter/tree/7ac581986811825ff10a86a37d8bd775ce9d6a8f/pb/src/entries))?
    - An example of a real hook I plan to write: an API endpoint that proxies HTTP requests to overcome CORS restrictions; basically [this](https://github.com/Leftium/proxy/blob/main/src/routes/p/%2Bserver.ts).
        - I suppose as a pocodex plugin, it would be simple to share/reuse in other projects?
2. Do you think a PocketBase/PocketHost [SvelteKit adapter](https://svelte.dev/docs/kit/writing-adapters) could be possible?
    - The adapter could output a pocodex plugin?
    - I think the main limiting factor would be if all required NodeJS API's and/or Kit dependencies work in the PocketBase JSVM
    - As a VM, could performance be another issue?

**Reply by benallfree** [reply]
Yes I’ve become fairly convinced that a TS code base that bundles all modules is the current best approach to JSVM. I’ve written two fairly large apps that way now, PocketPages being the biggest. 

Watch mode works great for me, I use it all the time. I actually use pocodex’s cli to watch and build. 

You could certainly break that code into a separate repo and make it a pocodex plugin. We need some contributors :) I’d be happy to help and also adapt pocodex for smoother integration. I hadn’t considered this use case but I like it. 

By sveltekit adapter I assume you mean SSR and other server-side endpoints? Since JSVM doesn’t have async, I think this would be really difficult and users would be constantly thwarted. I did start a little effort called PocketPorts that attempts to offer mostly-compatible synchronous alternatives to popular modules. For example, pocketbase-ejs does this for the popular ejs package. 

I also started PocketPages (pocketpages.dev) as an answer to robust JS-driven server pages and APIs. 

Gani will probably say it’s bad and that the JSVM performance is an order of magnitude slower, and he might be right, but I’m doing it anyway. With a good server-side JS framework I can ship so much faster. I can always move critical sections into Go if scaling becomes an issue.

**Reply by Leftium** [reply]
> By sveltekit adapter I assume you mean SSR and other server-side endpoints? Since JSVM doesn’t have async, I think this would be really difficult and users would be constantly thwarted.

That was my thinking, too. That's why I was figuring out how to do everything as a SPA (I usually prefer serverless apps).

But then I had two thoughts:
1. I thought `async` was just just syntactic sugar, and could transpile to promises. (But perhaps the JSVM can't do promises, either? I think I heard async Go is difficult...)
2. The pocketbase-jsvm docs looked like [async was now possible...](https://www.npmjs.com/package/pocketbase-jsvm#:~:text=//%20ECMAScript%202020%20features%20are%20supported%0Aconst%20asyncFunction%20%3D%20async%20()%20%3D%3E%20%7B%0A%20%20//%20Your%20code%20here%0A%7D)

**Reply by benallfree** [reply]
JSVM doesn't support promises either. Nothing async because there is no event loop. No setTimeout, etc.

Ooo, you found another error in the `pocketbase-jsvm` docs. I've updated it.

In a way, you don't need async because PocketBase runs multiple JS threads concurrently.

**Comment by VictorioBerra** [comment]
I tried to port pocodex TS stuff to my project and failed. I think im close but my ./dist folder is full of .cjs files which PB wont load (i think it looks for *.pb.js) and im not sure where this file is?

```ts
require("pocodex/dist/pb").Init();
```

https://github.com/benallfree/pocodex/blob/main/src/pb/pb_hooks/pocodex.pb.ts#L2

I dont get anything in a relative pocodex folder.
